### PR TITLE
Python: Fix iterable-unpacking taint CP

### PR DESCRIPTION
**Pull request by RasmusWL**
When running ql/python/ql/src/Security/CWE-079/ReflectedXss.ql against the database for flask.

Iitially there were 10 million result-tuples for iterable_unpacking_descent.

With this change, we're down to roughly 2100.

Fixes the problem introduced in #2700, and disabled in #2973

### CODE REVIEW

**Comment by RasmusWL** [comment]
Started https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/33/

**Comment by RasmusWL** [comment]
> I'm curious to see how the performance tests pan out.

https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/33/ looks fine. Are you asking for a dist-compare besides this?

> I would also like to see the RA generated/executed on a large snapshot for the predicates involved (i.e. the main predicate and the two helper predicates).

Alright. Is `saltstack/salt` large enough for you with ~600k lines of python code? and I think I need some guidance on _the RA generated/executed_ since I wouldn't know exactly which parts to pick from the query logs.

File: python/ql/src/semmle/python/dataflow/Implementation.qll [hunk 1]
```diff
@@ -967,6 +951,45 @@ private predicate piNodeTestAndUse(PyEdgeRefinement defn, ControlFlowNode test,
...
+        // base-case #0
+        depth = 0 and
+        result = parent_kind
+        or
+        // base-case #1
+        depth = 1 and
+        result = parent_kind.getMember()
+        or
+        // recursive case
+        result = taint_at_depth(parent_kind.getMember(), depth-1)
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/src/semmle/python/dataflow/Implementation.qll]: Fixed! This won't change the results, just the executions that's that we have to evaluate :+1:

**Comment by RasmusWL** [comment]
I ran `ReflectedXss.ql` on the current snapshot for saltstack/salt, and I didn't see any problems. (this I the same query I used doing my other tests)

I created a [gist](https://gist.github.com/RasmusWL/ecbb904840c99ccd9526cc3a89054a15) with the query logs, but that only shows the first 8 thousand lines, so you need to look at the [raw](https://gist.githubusercontent.com/RasmusWL/ecbb904840c99ccd9526cc3a89054a15/raw/2eecd243c6fbacc22b85e9002dae7036377e01dc/execute-queries-20200330.105407.841.log) results to see all 56k lines :wink: 

Looking at all the places it says `iterable_unpacking_descent` I didn't see anything of concern.

Is this sort of what you were asking for @tausbn? -- when you said:

> I would also like to see the RA generated/executed on a large snapshot for the predicates involved (i.e. the main predicate and the two helper predicates).

**Comment by tausbn** [comment]
> Alright. Is `saltstack/salt` large enough for you with ~600k lines of python code? and I think I need some guidance on _the RA generated/executed_ since I wouldn't know exactly which parts to pick from the query logs.

Sure, `salt` should be plenty big. What I meant by the RA would amount to:

1. Running an appropriate query, generating a log with tuple counts (since this has to do with taint tracking, I imagine most taint tracking queries would do).
2. Looking for the place in the log where the `iterable_unpacking_descent` and `taint_at_depth` predicates are evaluated, to see whether they result in a "reasonable" amount of tuples, whether there are any bad join orders or cartesian products, etc. Basically this amounts to searching for those names until you find "`tuple counts for `" followed by the name of the predicate.

For 1. you'll most likely want to use a recent version of VSCode, with the most recent improvements to log handling. (Or you could also just run it manually.)

**Comment by tausbn** [comment]
Oh, hah. I was in the process of writing my comment when you added yours. I'll take a quick look at the results.

**Pull request merged**